src/memory/pageTable.ts

```diff
@@ -0,0 +1,115 @@
+import assert from 'assert';
+
+export class PageTable {
+  public static readonly NUM_OF_ENTRIES: number = 512;
+  public static readonly ENTRY_SIZE: number = 8;
+  private static readonly EMPTY_FREE_LIST: number = -1;
+
+  private readonly memory: DataView;
+  private freeList: number;
+  private numOfFreeEntries: number;
+
+  constructor() {
+    const data = new ArrayBuffer(
+      PageTable.NUM_OF_ENTRIES * PageTable.ENTRY_SIZE
+    );
+    this.memory = new DataView(data);
+
+    // The free list is made of entries, where each free entry carries the offset of the next free entry
+    this.freeList = 0;
+    let i;
+    for (i = this.freeList; i <= PageTable.NUM_OF_ENTRIES; i++) {
+      this.memory.setFloat64(i * PageTable.ENTRY_SIZE, i + 1);
+    }
+    this.memory.setFloat64(i, PageTable.EMPTY_FREE_LIST);
+    this.numOfFreeEntries = PageTable.NUM_OF_ENTRIES;
+  }
+
+  /**
+   * Gets the data from the entry at the offset.
+   * Note: This should be the offset to an entry (9 bits),
+   * not an offset to the within the entry (12 bits)
+   */
+  public get(offset: number): number {
+    if (!this.isValidOffset(offset)) {
+      return -1;
+    }
+
+    return this.memory.getFloat64(offset * PageTable.ENTRY_SIZE);
+  }
+
+  /**
+   * Sets the first free entry in the free list to contain data.
+   */
+  public setFreeEntry(data: number): number {
+    const offset = this.freeList;
+
+    if (!this.isValidOffset(offset)) {
+      return -1;
+    }
+
+    this.freeList = this.get(offset);
+    this.memory.setFloat64(offset * PageTable.ENTRY_SIZE, data);
+    this.numOfFreeEntries -= 1;
+    return offset;
+  }
+
+  /**
+   * Updates the data contained in a used entry.
+   */
+  public setAllocatedEntry(offset: number, data: number): void {
+    if (!this.isValidOffset(offset)) {
+      return;
+    }
+
+    assert(!this.isFree(offset), 'Only allocated entries can be set.');
```

Comment:
Should we throw a `SegmentationFaultError` or equivalent instead?

Edit: I guess this is part of what you meant by handling errors properly.

Reply:
As discussed, segmentation faults should only be a concern when addressing unmapped virtual memory.